src/stock.rs

```diff
@@ -276,6 +272,1129 @@ impl StockMessage {
+                .unwrap_or_else(|| Cow::Borrowed(id.fallback())),
+        }
+    }
+}
+
+/// Stock string builder which allows retrieval of the message.
+///
+/// This builder allows retrieval of the message using [`TranslatedStockString::msg`], if it
+/// needs substitutions first however it provides further builder methods.
+struct TranslatedStockString<'a> {
+    context: &'a Context,
+    message: Cow<'static, str>,
+}
+
+impl<'a> TranslatedStockString<'a> {
+    /// Retrieves the built message.
+    fn msg(self) -> Cow<'static, str> {
+        self.message
+    }
+
+    /// Substitutes the first replacement value if one is present.
+    fn replace1(self, replacement: impl AsRef<str>) -> Self {
+        let msg = self
+            .message
+            .as_ref()
+            .replacen("%1$s", replacement.as_ref(), 1)
+            .replacen("%1$d", replacement.as_ref(), 1)
+            .replacen("%1$@", replacement.as_ref(), 1);
+        Self {
+            context: self.context,
+            message: Cow::Owned(msg),
+        }
+    }
+
+    /// Substitutes the second replacement value if one is present.
+    ///
+    /// Be aware you probably should have also called [`TranslatedStockString::replace1`] if
+    /// you are calling this.
+    fn replace2(self, replacement: impl AsRef<str>) -> Self {
+        let msg = self
+            .message
+            .as_ref()
+            .replacen("%2$s", replacement.as_ref(), 1)
+            .replacen("%2$d", replacement.as_ref(), 1)
+            .replacen("%2$@", replacement.as_ref(), 1);
+        Self {
+            context: self.context,
+            message: Cow::Owned(msg),
+        }
+    }
+
+    /// Augments the message by saying it was performed by a user.
+    ///
+    /// This looks up the display name of `contact` and uses the [`MsgActionByMe`] and
+    /// [`MsgActionByUser`] stock strings to turn the stock string in one that says the
+    /// action was performed by this user.
+    ///
+    /// E.g. this turns `Group image changed.` into `Group image changed by me.` or `Group
+    /// image changed by Alice`.
+    ///
+    /// Note that the original message should end in a `.`.
+    async fn action_by_contact(self, contact: u32) -> TranslatedStockString<'a> {
+        let message = self.message.as_ref().trim_end_matches('.');
+        let message = match contact {
+            DC_CONTACT_ID_SELF => MsgActionByMe::stock_str(self.context, message).await,
+            _ => {
+                let displayname = Contact::get_by_id(self.context, contact)
+                    .await
+                    .map(|contact| contact.get_name_n_addr())
+                    .unwrap_or_else(|_| format!("{}", contact));
+                MsgActionByUser::stock_str(self.context, message, displayname).await
+            }
+        };
+        TranslatedStockString {
+            context: self.context,
+            message,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) enum NoMessages {}
```

Comment:
Why `enum` and not `pub(crate) struct NoMessages;`?

Reply:
Not that this matter anymore, but a no-variant enum can not be constructed by anyone.  A no-fields struct **can** be constructed by anyone who can reach it (while a struct with private fields **can not** be constructed by anyone who can reach it).